src/protocols/bdx/BDXMessageUtils.cpp

```diff
@@ -0,0 +1,432 @@
+
+    if (hasDefLen)
+    {
+        if (widerange)
+        {
+            SuccessOrExit(bufReader.Read64(&aParsedMessage.mMaxLength).StatusCode());
+        }
+        else
+        {
+            SuccessOrExit(bufReader.Read32(&tmpUint32Value).StatusCode());
+            aParsedMessage.mMaxLength = tmpUint32Value;
+        }
+    }
+
+    SuccessOrExit(bufReader.Read16(&aParsedMessage.mFileDesLength).StatusCode());
+
+    // WARNING: this struct will store a pointer to the start of the file designator in the PacketBuffer,
+    // but will not make a copy. It is essential that this struct not outlive the PacketBuffer,
+    // or there is risk of unsafe memory access.
+    VerifyOrExit(bufReader.HasAtLeast(aParsedMessage.mFileDesLength), err = CHIP_ERROR_MESSAGE_INCOMPLETE);
+    aParsedMessage.mFileDesignator = &bufStart[bufReader.OctetsRead()];
+
+    // Rest of message is metadata (could be empty)
+    if (bufReader.Remaining() > aParsedMessage.mFileDesLength)
+    {
+        // WARNING: this struct will store a pointer to the start of metadata in the PacketBuffer,
+        // but will not make a copy. It is essential that this struct not outlive the PacketBuffer,
+        // or there is risk of unsafe memory access.
+        uint16_t metadataStartIndex    = static_cast<uint16_t>(bufReader.OctetsRead() + aParsedMessage.mFileDesLength);
+        aParsedMessage.mMetadata       = &bufStart[metadataStartIndex];
+        aParsedMessage.mMetadataLength = static_cast<uint16_t>(aBuffer.DataLength() - metadataStartIndex);
+    }
+
+exit:
+    if (bufReader.StatusCode() != CHIP_NO_ERROR)
+    {
+        err = bufReader.StatusCode();
+    }
+    return err;
+}
+
+size_t TransferInit::PackedSize() const
+{
+    bool widerange    = ((mStartOffset | mMaxLength) > std::numeric_limits<uint32_t>::max());
+    size_t offsetSize = widerange ? 8 : 4;
+    size_t lengthSize = offsetSize;
+
+    // First 2 bytes are Transfer Control and Range Control bytes
+    return (2 + sizeof(mMaxBlockSize) + offsetSize + lengthSize + sizeof(mFileDesLength) + mFileDesLength + mMetadataLength);
+}
+
+bool TransferInit::operator==(const TransferInit & another) const
+{
+    bool fileDesMatches  = memcmp(mFileDesignator, another.mFileDesignator, mFileDesLength) == 0;
+    bool metadataMatches = memcmp(mMetadata, another.mMetadata, mMetadataLength) == 0;
+
+    return (mSupportedVersions == another.mSupportedVersions && mSupportsSenderDrive == another.mSupportsSenderDrive &&
+            mSupportsReceiverDrive == another.mSupportsReceiverDrive && mSupportsAsync == another.mSupportsAsync &&
+            mStartOffset == another.mStartOffset && mMaxLength == another.mMaxLength && mMaxBlockSize == another.mMaxBlockSize &&
+            fileDesMatches && metadataMatches);
+}
+
+CHIP_ERROR SendAccept::Pack(System::PacketBuffer & aBuffer) const
+{
+    CHIP_ERROR err      = CHIP_NO_ERROR;
+    uint8_t transferCtl = 0;
+
+    BufBound bbuf(aBuffer.Start(), aBuffer.AvailableDataLength());
```

Comment:
Is there a test in PacketBuffer to ensure that `AvailableDataLength()` matches reality?

The chain PacketBuffer::AvailableDataLength() -> PacketBuffer::MaxDataLengths() -> PacketBuffer::AllocSize() refers to `inline uint16_t PacketBuffer::AllocSize() const` which has different logic from `lAllocSize/lBlockSize` in PacketBuffer::NewWithAvailableSize().

@andreilitvin @rwalker-apple I think the above should carefully be double-checked/tested across platforms as I am realizing that making assumptions about "AvailableDataLength()" seems to rely on a huge chain of arithmetic that is quite opaque.

Reply:
Just want to add that there is a test "CheckAvailableDataLength" in `TestSystemPacketBuffer.cpp`.